src/metabase/sfc/classify.clj

```diff
@@ -0,0 +1,198 @@
+
+(def ^:private ^:const ^Float percent-valid-url-threshold
+  "Fields that have at least this percent of values that are valid URLs should be given a special type of `:type/URL`."
+  0.95)
+
+(def ^:const ^Integer low-cardinality-threshold
+  "Fields with less than this many distinct values should automatically be given a special type of `:type/Category`."
+  300)
+
+(def ^:private ^:const ^Integer average-length-no-preview-threshold
+  "Fields whose values' average length is greater than this amount should be marked as `preview_display = false`."
+  50)
+
+;; TODO - we should make all these `test:` functions private since they're only used in this namespace.
+;; In order to do that we need to tweak the corresponding tests however
+(defn test:category-type
+  "When no initial guess of the special type, based on the fields name, was found
+   and the field has less than `low-cardinality-threshold`
+ default to :type/Category"
+  ;; this used to only apply to new fields and that was removed in refactor, does that break things
+  [{:keys [base_type visibility_type name is_fk is_pk] :as fingerprint} {:keys [special-type] :as field-stats}]
+  (if (and (not is_fk) (not is_pk )
+           (nil? (:special-type field-stats))
+           (< 0 (:cardinality fingerprint) low-cardinality-threshold)
+           (field-values/field-should-have-field-values? {:base_type       base_type
+                                                          :special_type    special-type
+                                                          :visibility_type visibility_type
+                                                          :name            name}))
+    (assoc field-stats :special-type :type/Category)
+    field-stats))
+
+(defn test:no-preview-display
+  "If FIELD's is textual and its average length is too great, mark it so it isn't displayed in the UI."
+  [fingerprint field-stats]
+  (if-not (and (= :normal (:visibility_type fingerprint))
+               (isa? (:base_type fingerprint) :type/Text))
+    ;; this field isn't suited for this test
+    field-stats
+    ;; test for avg length
+    (let [avg-len (:field_avg_length fingerprint)]
+      (if-not (and avg-len (> avg-len average-length-no-preview-threshold))
+        field-stats
+        (do
+          (log/debug (u/format-color 'green "Field '%s' has an average length of %s. Not displaying it in previews." (:qualified-name fingerprint) avg-len))
+          (assoc field-stats :preview-display false))))))
+
+(defn test:url-special-type
+  "If FIELD is texual, doesn't have a `special-type`, and its non-nil values are primarily URLs, mark it as `special-type` `:type/URL`."
+  [fingerprint field-stats]
+  (if-not (and (not (:special-type field-stats))
+               (isa? (:base_type fingerprint) :type/Text))
+    ;; this field isn't suited for this test
+    field-stats
+    ;; test for url values
+    (let [percent-urls (:field-percent-urls fingerprint)]
```

Comment:
@salsakran do you have a sense if any of this code ever worked? I keep stumbling upon new things like this when I read the code. Here we're trying to look up the key `:field-percent-urls` in the fingerprint. However, the fingerprint uses the key `:field_percent_urls`